source/menu/m_menu.qc

```diff
@@ -550,6 +552,146 @@ void(float order, vector minmaxsteps, string cvar_s, float is_int, float no_text
 	}
 };
 
+static void(string id, vector pos, vector size, __inout string text, __inout float cursor) handle_text_input =
+{
+	float maxlen = 0;
+
+	if (getTextWidth(text, MENU_TEXT_SMALL.x) + 18 >= size.x)
+		maxlen = strlen(text);
+	else
+		maxlen = 128;
+
+	sui_action_element(pos, size, id, sui_noop);
+	if (sui_is_clicked(id)) cursor = strlen(text);
+	if (sui_is_hovered(id))
+	{
+		float char = 0;
+		float scan = 0;
+		while(sui_get_input(char, scan)) sui_handle_text_input(char, scan, maxlen, text, cursor);
+	}
+};
+
+static void(string id, vector pos, vector size, string text, float cursor) draw_text_input =
+{
+	sui_push_frame(pos - [6, 5], size);
+	sui_fill([0, 0], size, [0, 0, 0], 1.0, 0);
+	sui_border_box([0, 0], size, 2, [0.2, 0.2, 0.2], 0.3, 0);
+	sui_set_align([SUI_ALIGN_START, SUI_ALIGN_CENTER]);
+	sui_text([6, 0], MENU_TEXT_SMALL, text, MENU_TEXT_1, 1, 0);
+	float cursor_x = getTextWidth(substring(text, 0, cursor), MENU_TEXT_SMALL.x);
+	if (sui_is_hovered(id))
+		sui_text([6 + cursor_x, 0], MENU_TEXT_SMALL, "|", [1.0, 0.25, 0.25], 1, 0);
+	sui_pop_frame();
+};
+
+//
+// Menu_TextInput(id, order, text, cursor)
+// Draws a Text Input box at the given order position
+//
+void(string id, float order, __inout string text, __inout float cursor) Menu_TextInput =
+{
+	vector pos = [320, 50 + (order * 25)];
+	vector size = [290, 22];
+
+	handle_text_input(id, pos, size, text, cursor);
+
+	draw_text_input(id, pos, size, text, cursor);
+};
+
+//
+// Menu_PasswordInput(id, order, text, cursor)
+// Draws a Text Input box at the given order position (blocked out with asterisks)
+//
+void(string id, float order, __inout string text, __inout float cursor) Menu_PasswordInput =
+{
+	vector pos = [320, 50 + (order * 25)];
+	vector size = [290, 22];
+	string safe = "";
+
+	for (int i = 0; i < strlen(text); i++)
+		safe = sprintf("%s%s", safe, "*");
+
+	handle_text_input(id, pos, size, text, cursor);
+
+	draw_text_input(id, pos, size, safe, cursor);
+};
+
+static string(string s) str2ascii =
+{
+	for (int i = 0; i < strlen(s); i++)
+	{
+		if (s[i] < 32 || s[i] > 126)
+			s[i] = 63; // question mark
```

Comment:
> * We should also filter out non-ascii characters since our fontmap doesn't support them 😞

this converts the string to a safe ASCII range, as requested

Reply:
This was directed to the comment left. Is 63 ascii for `?`? Or is this an expression of confusion?

Reply:
`63` is the ASCII value for `?`, correct